### PR TITLE
bpf: Only create conntrack entries for SYN packets

### PR DESCRIPTION
**Pull request by joestringer**
Previously in e4c58da6d6a5 ("bpf: Simplify connection tracking logic"),
we loosened the policy for creating conntrack entries to allow entries
for long-lived connections to be temporarily removed and to be
subsequently added. This worked around issues with overzealous conntrack
cleanup and proxy implementations with no keepalives. After recent
commits to simplify & streamline conntrack garbage collection, and
updates to the keepalive policy, we expect that it is now safe again to
require a SYN packet to create a connection.

This also allows deletion of conntrack entries due to commit 9095ef8918e3
("bpf: Remove connection tracking entries on policy deny") to halt
ongoing connections when a liberal egress policy is in place, rather than
allowing such connections to continue due to egress traffic re-creating
the conntrack entry.

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

File: bpf/lib/conntrack.h [hunk 1]
```diff
@@ -231,10 +231,10 @@ static inline int __inline__ ct_lookup6(void *map, struct ipv6_ct_tuple *tuple,
 			if (skb_load_bytes(skb, l4_off + 12, &flags, 2) < 0)
 				return DROP_CT_INVALID_HDR;
 
-			if (unlikely(flags.rst || flags.fin))
-				action = ACTION_CLOSE;
-			else
+			if (unlikely(flags.syn && !flags.ack))
```
Comment (manalibhutiyani) [inline bpf/lib/conntrack.h]: Just curious: Should this be `if (likely(flags.syn && !flags.ack))` ? Not sure what's the likelihood of SYN occurring :)

**Comment by tgraf** [comment]
We previously had a check to actually drop such packets:
```
-       /* No entries found, packet must be eligible for creating a CT entry */
-       if (ret == CT_NEW && action != ACTION_CREATE)
-               ret = DROP_CT_CANT_CREATE;
-
```

We either re-add this check or we have ct_lookup indicate to the caller when to create it, right now, creation happens on return of `CT_NEW`:

```
        switch (ret) {
        case CT_NEW:
                /* New connection implies that rev_nat_index remains untouched
                 * to the index provided by the loadbalancer (if it applied).
                 * Create a CT entry which allows to track replies and to
                 * reverse NAT.
                 */
                ct_state_new.src_sec_id = SECLABEL;
                ret = ct_create6(&CT_MAP6, tuple, skb, CT_EGRESS, &ct_state_new);
                if (IS_ERR(ret))
                        return ret;
                break;
```

File: bpf/lib/conntrack.h (hunk 1 above)
Comment (jrajahalme) [reply bpf/lib/conntrack.h]: SYN is set and ACK is not set only on the very first packet a TCP connection, so only a fraction of TCP packets has this set of flags, hence the unlikely().

Comment (manalibhutiyani) [reply bpf/lib/conntrack.h]: @jrajahalme Thanks for the explanation!

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Looking at the log of the commit I added a revert for, I'm a little nervous this may reintroduce #2752 which we don't appear to have seen in the past two months.

File: bpf/lib/conntrack.h [hunk 2]
```diff
@@ -231,10 +231,10 @@ static inline int __inline__ ct_lookup6(void *map, struct ipv6_ct_tuple *tuple,
...
 				action = ACTION_CREATE;
+			else
+				action = ACTION_CLOSE;
```
Comment (jrajahalme) [inline bpf/lib/conntrack.h]: If the intent is to not allow creation of the entry for non-SYN/nACK packets, why are we changing the closing logic at the same time? How about this:
```
if (unlikely(flags.syn && !flags.ack))
  action = ACTION_CREATE;
else if (unlikely(flags.rst || flags.fin))
  action = ACTION_CLOSE;
/* else leave action as ACTION_UNSPEC */
```

**Review by jrajahalme — DISMISSED** [review]
We should not change the case where we set `action = ACTION_CLOSE`, i.e., the change should be limited to when se set `action = ACTION_CREATE`.

**Comment by tgraf** [comment]
> Looking at the log of the commit I added a revert for, I'm a little nervous this may reintroduce #2752 which we don't appear to have seen in the past two months.

Wouldn't we already drop these packets ever since we moved the proxy port into the policy map?

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: bpf/lib/conntrack.h (hunk 2 above)
Comment (joestringer) [reply bpf/lib/conntrack.h]: I did exactly this, thanks.

**Comment by joestringer** [comment]
@tgraf I think that that case involved the proxy closing a connection when the conntrack entry had already been deleted, and the connection failed to properly close it because we wouldn't won't let the FIN through. Maybe we're a bit better behaved now that there's an ACTION_CLOSE which lets RST/FIN connections linger a little longer.

**Review by jrajahalme — APPROVED** [review]

**Review by tgraf — APPROVED** [review]
Only needs rebase

**Comment by joestringer** [comment]
@tgraf rebased.

**Comment by joestringer** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by joestringer** [comment]
Removing `stable/needs-backport` as we've had reports of problems introduced by this commit.

**Pull request merged**